packages/react-aria-components/src/GridList.tsx

```diff
@@ -47,19 +76,127 @@ function GridList<T extends object>(props: GridListProps<T>, ref: ForwardedRef<H
+  let dragPreview: JSX.Element;
+  let preview = useRef(null);
+
+  if (isListDraggable) {
+    dragState = dragAndDropHooks.useDraggableCollectionState({
+      collection,
+      selectionManager,
+      preview: dragAndDropHooks.renderDragPreview ? preview : null
+    });
+    dragAndDropHooks.useDraggableCollection({}, dragState, ref);
+
+    dragPreview = dragAndDropHooks.renderDragPreview
+      ? <dragAndDropHooks.DragPreview ref={preview}>{dragAndDropHooks.renderDragPreview}</dragAndDropHooks.DragPreview>
+      : null;
+  }
+
+  if (isListDroppable) {
+    dropState = dragAndDropHooks.useDroppableCollectionState({
+      collection,
+      selectionManager
+    });
+
+    let keyboardDelegate = new ListKeyboardDelegate(
+      collection,
+      selectionManager.disabledBehavior === 'selection' ? new Set() : selectionManager.disabledKeys,
+      ref
+    );
+    let dropTargetDelegate = dragAndDropHooks.dropTargetDelegate || new dragAndDropHooks.ListDropTargetDelegate(collection, ref);
+    droppableCollection = dragAndDropHooks.useDroppableCollection({
+      keyboardDelegate,
+      dropTargetDelegate
+    }, dropState, ref);
+
+    isRootDropTarget = dropState.isDropTarget({type: 'root'});
+  }
+
+  let renderProps = useRenderProps({
+    className: props.className,
+    style: props.style,
+    defaultClassName: 'react-aria-GridList',
+    values: {
+      isDropTarget: isRootDropTarget,
+      isEmpty: state.collection.size === 0
+    }
+  });
+
+  let emptyState;
+  let emptyStatePropOverrides;
+  if (state.collection.size === 0 && props.renderEmptyState) {
+    // Ideally we'd use `display: contents` on the row and cell elements so that
+    // they don't affect the layout of the children. However, WebKit currently has
+    // a bug that makes grid elements with display: contents hidden to screen readers.
+    // https://bugs.webkit.org/show_bug.cgi?id=239479
+    let content = props.renderEmptyState();
+    if (isWebKit()) {
+      // For now, when in an empty state, swap the role to group in webkit.
+      emptyStatePropOverrides = {
+        role: 'group',
+        'aria-multiselectable': undefined
+      };
+
+      if (isIOS()) {
+        // iOS VoiceOver also doesn't announce the aria-label of group elements
+        // so try to add a visually hidden label element as well.
+        emptyState = (
+          <>
+            <VisuallyHidden>{gridProps['aria-label']}</VisuallyHidden>
+            {content}
+          </>
+        );
+      } else {
+        emptyState = content;
+      }
+    } else {
+      emptyState = (
+        <div role="row" style={{display: 'contents'}}>
+          <div role="gridcell" style={{display: 'contents'}}>
+            {content}
+          </div>
+        </div>
+      );
+    }
+  }
+
   return (
-    <ul
+    <div
       {...filterDOMProps(props)}
-      {...gridProps}
+      {...renderProps}
+      {...mergeProps(gridProps, droppableCollection?.collectionProps, emptyStatePropOverrides)}
```

Comment:
why aren't we merging all of these?

Reply:
useRenderProps only ever returns `className`, `style`, and `children` which can't come from the others so same reason we don't merge those props in RSP?